### PR TITLE
Close #2800 Update Global Footer with new X/Twitter Icon.

### LINKED ISSUE #2800: Update Global Footer with new Twitter Icon
Twitter updated their icon, https://github.com/az-digital/az-icons/pull/38

Need to update demo content with new icon class .`az-icon-x-twitter`

https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_global_footer/data/az_global_footer.json#L393-L406

Create a database update to change the class for existing sites similar to this one 
https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_global_footer/az_global_footer.install#L48-L79

Shouldn't need to update the library, since it is using the CDN and is using the main branch.
https://github.com/az-digital/az_quickstart/blob/main/themes/custom/az_barrio/includes/common.inc#L11

### PR DESCRIPTION
**Pull request by mrbeiley**
## Description

Given Twitter's rebranding to X, update Global Footer with new 'X' Icon. Add a function to update link(s) & icon(s) in the global footer.

## Related issues

Close Issue #2800 

## How to test

- Install Global Footer module
- View changes to the X/Twitter name and logo

- lando local development: `lando start` `lando phpunit`

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [x] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by danahertzberg** [comment]
Need to test on Pantheon multidev

File: modules/custom/az_global_footer/az_global_footer.install [hunk 1]
```diff
@@ -139,3 +139,36 @@ function az_global_footer_update_1020703() {
...
   return t('Updated %count menu link(s) from A-Z Index to Directory in the global footer.', ['%count' => $updated_count]);
 }
+
+/**
+ * Update Twitter link(s) to new X link(s)
+ *
+ * Twitter icon(s) in the global footer will be updated with
+ * new X icon(s).
+ */
+function az_global_footer_update_1020704() {
```
Comment (mrbeiley) [inline modules/custom/az_global_footer/az_global_footer.install:149]: Update the function name; needs to reflect patch release

File: modules/custom/az_global_footer/data/az_global_footer.json [hunk 2]
```diff
@@ -393,14 +393,14 @@
       "link_id": "36",
       "parent_link_id": "0",
       "menu": "az-footer-social-media",
-      "title": "Twitter",
-      "urlpath": "https://twitter.com/uarizona",
+      "title": "X-Twitter",
```
Comment (danahertzberg) [inline modules/custom/az_global_footer/data/az_global_footer.json:396]: Need to verify this title with central Marcomm

File: modules/custom/az_global_footer/az_global_footer.install (hunk 1 above)
Comment (danahertzberg) [reply modules/custom/az_global_footer/az_global_footer.install:149]: https://github.com/az-digital/az_quickstart/blob/main/RELEASES.md#update-hook-numbering-convention

File: modules/custom/az_global_footer/data/az_global_footer.json (hunk 2 above)
Comment (trackleft) [reply modules/custom/az_global_footer/data/az_global_footer.json:396]: Has this been verified?

File: modules/custom/az_global_footer/az_global_footer.install (hunk 1 above)
Comment (trackleft) [reply modules/custom/az_global_footer/az_global_footer.install:149]: looks like it needs to be 1020801

Comment (trackleft) [inline modules/custom/az_global_footer/az_global_footer.install:149]:
```suggestion
function az_global_footer_update_1020801() {
```

**Comment by trackleft** [comment]
Seems to work 
image

File: modules/custom/az_global_footer/data/az_global_footer.json (hunk 2 above)
Comment (mrbeiley) [reply modules/custom/az_global_footer/data/az_global_footer.json:396]: Henry Greenberg suggested "X, formerly known as Twitter". Gonna make the change now.

**Review by trackleft — DISMISSED** [review]

**Comment by joshuasosa** [comment]
IMO "X, formerly known as Twitter" looks too wordy and doesn't look good.
image

https://x.com/uarizona redirects to https://twitter.com/uarizona

Until it's the other way around, where Twitter redirects everything to X, I don't think it should be changed. It's an unnecessary redirect, though I do get wanting to get ahead of the ball and not update again later.

But it's eventually going to need to be updated later anyways, to get rid of any reference to Twitter when it's completely gone.

Until it's all said and gone and all things point to X, I'd suggest:
Label: Twitter (X)
URL: https://twitter.com/uarizona

**Review by danahertzberg — DISMISSED** [review]

**Comment by danahertzberg** [comment]
> IMO "X, formerly known as Twitter" looks too wordy and doesn't look good. 
> 
> Until it's all said and gone and all things point to X, I'd suggest: Label: Twitter (X) URL: https://twitter.com/uarizona

I don't mind the redirect to stay ahead of things. The "X, formerly known as Twitter" link title is a bit wordy, but we can thank Twitter/X for that lol. @HenryzGreenberg @mrbeiley could you give a final opinion on this link text?

**Comment by HenryzGreenberg** [comment]
The Associated Press updated its stylebook, which many regard as the gold standard, to reflect the name change. It suggested that the company and its social media platform be called “X, formerly known as Twitter, so that is what I would suggest. However that ends up looking awful. [https://app.screencast.com/eR23LhKVRro3B](https://app.screencast.com/eR23LhKVRro3B ) so we opted for the truncated version "X, formerly Twitter" 
[https://app.screencast.com/LXFcnMUMv9oAh](https://app.screencast.com/LXFcnMUMv9oAh)

File: modules/custom/az_global_footer/data/az_global_footer.json [hunk 3]
```diff
@@ -393,14 +393,14 @@
       "link_id": "36",
       "parent_link_id": "0",
       "menu": "az-footer-social-media",
-      "title": "Twitter",
-      "urlpath": "https://twitter.com/uarizona",
+      "title": "X, formerly known as Twitter",
```
Comment (joeparsons) [inline modules/custom/az_global_footer/data/az_global_footer.json:396]:
```suggestion
      "title": "X, formerly Twitter",
```
Per @HenryzGreenberg

File: modules/custom/az_global_footer/az_global_footer.install [hunk 4]
```diff
@@ -139,3 +139,36 @@ function az_global_footer_update_1020703() {
...
+    ->accessCheck(FALSE)
+    ->condition('menu_name', 'az-footer-resources')
+    ->condition('title', 'X')
+    ->condition('link__uri', 'https://x.com/uarizona')
+    ->execute();
+
+  foreach ($menu_link_content_ids as $id) {
+    /**
+     * @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content
+     */
```
Comment (joeparsons) [inline modules/custom/az_global_footer/az_global_footer.install:161]: Nitpick:
```suggestion
    /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
```

**Comment by joshuasosa** [comment]
I don't see why AP needs to be religiously followed for every little thing. They are not end all be all for all formatting unless it's a UA technical policy that I missed. Their format here does not look good nor does it increase reader comprehension.

But just an opinion against at this point since it doesn't affect me.

**Comment by danahertzberg** [comment]
> I don't see why AP needs to be religiously followed for every little thing. They are not end all be all for all formatting unless it's a UA technical policy that I missed. Their format here does not look good nor does it increase reader comprehension.
> 
> But just an opinion against at this point since it doesn't affect me.

AP style is religiously followed by UA as far as I know. The [written style guide also indicates using the AP Stylebook for any usages not addressed](https://marcom.arizona.edu/communications/written-style-guide#:~:text=For%20guidance%20on%20usages%20not%20addressed%20here%2C%20please%20consult%20the%20AP%20Stylebook) on the brand site, like Twitter.

I like the "X, formerly Twitter" stance and think we should go with it per guidance from University Marketing & Communications.

**Review by danahertzberg — DISMISSED** [review]

**Review by sanjanans — APPROVED** [review]

File: modules/custom/az_global_footer/az_global_footer.install [hunk 5]
```diff
@@ -139,3 +139,34 @@ function az_global_footer_update_1020703() {
...
+    ->condition('title', 'X')
+    ->condition('link__uri', 'https://x.com/uarizona')
+    ->execute();
+
+  foreach ($menu_link_content_ids as $id) {
+    /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
+    $menu_link_content = MenuLinkContent::load($id);
+    if ($menu_link_content !== NULL) {
+      // phpcs:disable
+      /* @phpstan-ignore-next-line */
```
Comment (danahertzberg) [inline modules/custom/az_global_footer/az_global_footer.install:163]: Investigate why these lines are needed before merging @tadean @bberndt-uaz

**Review by camikazegreen — DISMISSED** [review]
Happy to let MarCom make final determination on wording and display. Doesn't wrap on the most common breakpoints, so it looks ok to me.

File: modules/custom/az_global_footer/az_global_footer.install (hunk 5 above)
Comment (bberndt-uaz) [reply modules/custom/az_global_footer/az_global_footer.install:163]: The PHPStan error suppressed is:
```
Access to an undefined property Drupal\Core\Field\FieldItemList::$uri.                      
     💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property
```
The PHPCS warning suppressed is: 
```
Filesystem function link used but not as a function
```

Since we have used the same code to update other Global Footer links (see the previous update functions in this file), I think we can go ahead and merge this without changing these lines.

File: modules/custom/az_global_footer/az_global_footer.install [hunk 6]
```diff
@@ -139,3 +139,34 @@ function az_global_footer_update_1020703() {
...
+      $menu_link_content->link->uri = 'https://x.com/uarizona';
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:164]: I believe the function needs to match the correct link to update and then make the changes to the title, the URI, and the icon classes. **This code needs to be tested before merging.**

```suggestion
    ->condition('menu_name', 'az-footer-social-media')
    ->condition('title', 'Twitter')
    ->condition('link__uri', 'https://twitter.com/uarizona')
    ->execute();

  foreach ($menu_link_content_ids as $id) {
    /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
    $menu_link_content = MenuLinkContent::load($id);
    if ($menu_link_content !== NULL) {
      // phpcs:disable
      /* @phpstan-ignore-next-line */
      $menu_link_content->title = 'X, formerly Twitter';
      /* @phpstan-ignore-next-line */
      $menu_link_content->link->uri = 'https://x.com/uarizona';
      /* @phpstan-ignore-next-line */
      $menu_link_content->link->options->attributes->class = 'az-icon-twitter az-icon-x-twitter az-icon-spacing';
```

**Review by bberndt-uaz — CHANGES_REQUESTED** [review]

File: modules/custom/az_global_footer/az_global_footer.install (hunk 6 above)
Comment (trackleft) [reply modules/custom/az_global_footer/az_global_footer.install:164]: Great catch @bberndt-uaz

Comment (trackleft) [reply modules/custom/az_global_footer/az_global_footer.install:164]: I'll merge and apply to a multidev

**Comment by trackleft** [comment]
Now we are getting this 
image

File: modules/custom/az_global_footer/az_global_footer.install [hunk 7]
```diff
@@ -139,3 +139,38 @@ function az_global_footer_update_1020703() {
...
+    /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
+    $menu_link_content = MenuLinkContent::load($id);
+    if ($menu_link_content !== NULL) {
+      // phpcs:disable
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->title = 'X, formerly Twitter';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->uri = 'https://x.com/uarizona';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->options->attributes->class = 'az-icon-twitter az-icon-x-twitter az-icon-spacing';
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:168]:
```suggestion
      $menu_link_content->link->options['attributes']['class'] = 'az-icon-twitter az-icon-x-twitter az-icon-spacing';
```

File: modules/custom/az_global_footer/az_global_footer.install [hunk 8]
```diff
@@ -139,3 +139,38 @@ function az_global_footer_update_1020703() {
...
+    /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
+    $menu_link_content = MenuLinkContent::load($id);
+    if ($menu_link_content !== NULL) {
+      // phpcs:disable
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->title = 'X, formerly Twitter';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->uri = 'https://x.com/uarizona';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->options['attributes']['class'] = 'az-icon-twitter az-icon-x-twitter az-icon-spacing';
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:168]:
```suggestion
      $menu_link_content->link->options['attributes']['class'] = [
        'az-icon-twitter',
        'az-icon-x-twitter',
        'az-icon-spacing'
        ];
```

File: modules/custom/az_global_footer/az_global_footer.install [hunk 9]
```diff
@@ -139,3 +139,42 @@ function az_global_footer_update_1020703() {
...
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->title = 'X, formerly Twitter';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->uri = 'https://x.com/uarizona';
+      /* @phpstan-ignore-next-line */
+      $menu_link_content->link->options['attributes']['class'] = [
+        'az-icon-twitter',
+        'az-icon-x-twitter',
+        'az-icon-spacing'
+        ];
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:172]:
```suggestion
      $menu_link_content->link->first()->options['attributes']['class'] = [
        'az-icon-twitter',
        'az-icon-x-twitter',
        'az-icon-spacing'
        ];
```

File: modules/custom/az_global_footer/az_global_footer.install [hunk 10]
```diff
@@ -139,3 +139,42 @@ function az_global_footer_update_1020703() {
...
+      // phpcs:enable
+      $menu_link_content->save();
+      $updated_count++;
+    }
+  }
+
+  return t('Updated %count X-Twitter link(s) & icon(s) in the global footer.', ['%count' => $updated_count]);
+}
```
Comment (trackleft) [inline modules/custom/az_global_footer/az_global_footer.install:180]:
```suggestion

/**
 * Update Twitter link(s) to new X link(s)
 *
 * Twitter icon(s) in the global footer will be updated with
 * new X icon(s).
 */
function az_global_footer_update_1020811() {
    $updated_count = 0;
    $menu_link_content_ids = \Drupal::entityQuery('menu_link_content')
        ->accessCheck(FALSE)
        ->condition('menu_name', 'az-footer-social-media')
        ->condition('title', 'X, formerly Twitter')
        ->condition('link__uri', 'https://x.com/uarizona')
        ->execute();

    foreach ($menu_link_content_ids as $id) {
        /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
        $menu_link_content = MenuLinkContent::load($id);

        if ($menu_link_content !== NULL) {
            // phpcs:disable
            /* @phpstan-ignore-next-line */
            $menu_link_content->title = 'X, formerly Twitter';
            /* @phpstan-ignore-next-line */
            $menu_link_content->link->uri = 'https://x.com/uarizona';
            /* @phpstan-ignore-next-line */

            // Retrieve existing options
            $options = $menu_link_content->link->first()->options;

            // Set new classes
            $options['attributes']['class'] = [
                'az-icon-x-twitter',
                'az-icon-spacing'
            ];

            // Update the options
            $menu_link_content->link->first()->options = $options;

            // phpcs:enable
            $menu_link_content->save();
            $updated_count++;
        }
    }
    return t('Updated %count X-Twitter link(s) & icon(s) in the global footer.', ['%count' => $updated_count]);
}
```

File: modules/custom/az_global_footer/az_global_footer.install [hunk 11]
```diff
@@ -139,3 +139,42 @@ function az_global_footer_update_1020703() {
...
+      $menu_link_content->link->first()->options['attributes']['class'] = [
+        'az-icon-twitter',
+        'az-icon-x-twitter',
+        'az-icon-spacing'
+        ];
+      // phpcs:enable
+      $menu_link_content->save();
+      $updated_count++;
+    }
+  }
```
Comment (trackleft) [inline modules/custom/az_global_footer/az_global_footer.install:177]:
```suggestion
    foreach ($menu_link_content_ids as $id) {
        /** @var \Drupal\menu_link_content\Entity\MenuLinkContent $menu_link_content */
        $menu_link_content = MenuLinkContent::load($id);

        if ($menu_link_content !== NULL) {
            // phpcs:disable
            /* @phpstan-ignore-next-line */
            $menu_link_content->title = 'X, formerly Twitter';
            /* @phpstan-ignore-next-line */
            $menu_link_content->link->uri = 'https://x.com/uarizona';
            /* @phpstan-ignore-next-line */

            // Retrieve existing options
            $options = $menu_link_content->link->first()->options;

            // Set new classes
            $options['attributes']['class'] = [
                'az-icon-x-twitter',
                'az-icon-spacing'
            ];

            // Update the options
            $menu_link_content->link->first()->options = $options;

            // phpcs:enable
            $menu_link_content->save();
            $updated_count++;
        }
    }
```

File: modules/custom/az_global_footer/az_global_footer.install [hunk 12]
```diff
@@ -139,3 +139,51 @@ function az_global_footer_update_1020703() {
...
+            $options = $menu_link_content->link->first()->options;
+
+            // Set new classes
+            $options['attributes']['class'] = [
+                'az-icon-x-twitter',
+                'az-icon-spacing'
+            ];
+
+            // Update the options
+            $menu_link_content->link->first()->options = $options;
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:180]:
```suggestion

            // Retrieve existing options
            /* @phpstan-ignore-next-line */
            $options = $menu_link_content->link->first()->options;

            // Set new classes
            $options['attributes']['class'] = [
                'az-icon-x-twitter',
                'az-icon-spacing'
            ];

            // Update the options
            /* @phpstan-ignore-next-line */
            $menu_link_content->link->first()->options = $options;
```

**Review by trackleft — DISMISSED** [review]

File: modules/custom/az_global_footer/az_global_footer.install [hunk 13]
```diff
@@ -139,3 +139,52 @@ function az_global_footer_update_1020703() {
...
+            $options['attributes']['class'] = [
+            'az-icon-x-twitter',
+            'az-icon-spacing'
+        ];
+
+            // Update the options
+            /* @phpstan-ignore-next-line */
+            $menu_link_content->link->first()->options = $options;
+
+            // phpcs:enable
```
Comment (bberndt-uaz) [inline modules/custom/az_global_footer/az_global_footer.install:183]:
```suggestion
      // phpcs:disable
      /* @phpstan-ignore-next-line */
      $menu_link_content->title = 'X, formerly Twitter';
      /* @phpstan-ignore-next-line */
      $menu_link_content->link->uri = 'https://x.com/uarizona';

      // Retrieve existing options
      /* @phpstan-ignore-next-line */
      $options = $menu_link_content->link->first()->options;

      // Set new classes
      $options['attributes']['class'] = [
        'az-icon-x-twitter',
        'az-icon-spacing'
      ];

      // Update the options
      /* @phpstan-ignore-next-line */
      $menu_link_content->link->first()->options = $options;

      // phpcs:enable
```

**Review by trackleft — APPROVED** [review]

**Review by bberndt-uaz — APPROVED** [review]
Thanks for all your help with the database update @trackleft and @tadean!

**Review by tadean — APPROVED** [review]
